chroma-common/blockdevices/blockdevice_zfs.py

```diff
@@ -129,8 +124,19 @@ def lock_pool(self):
         while not self.lock.i_am_locking():
             try:
                 self.lock.acquire(self.LOCK_ACQUIRE_TIMEOUT)
+                with open(self.lock.lock_file, 'w') as f:
+                    f.writelines(str(self.lock.pid))
             except LockTimeout:
-                pass
+                # prune locks owned by non-existent processes
+                with open(self.lock.lock_file, 'r') as f:
+                    contents = f.readlines()
+
+                assert len(contents) == 1 and contents[0].isdigit(), \
+                    'unexpected contents of lockfile %s: %s' % (self.lock.lock_file, contents)
+
+                # validate pid holding lock exists
+                if not pid_exists(int(contents[0])):
```

Comment:
it is the process that is doing the locking that is terminated, and in a utopia these kind of approaches would not be necessary, however in the real world storing PIDs in a lockfile and ignoring/removing lockfiles with PIDs of non-existent processes is a frequently used approach in unix. I experimented with various approaches including clearing up lockfiles on process termination, in the end this was the most effective solution, removing a lock if it is not held by an active process.